libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp

```diff
@@ -122,26 +365,283 @@ namespace hpx { namespace parallel { inline namespace v1 {
+namespace hpx { namespace parallel { inline namespace v1 {
+
+    // clang-format off
+    template <typename ExPolicy, typename Rng,
+        HPX_CONCEPT_REQUIRES_(hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_range<Rng>::value)>
+    // clang-format on
+    HPX_DEPRECATED_V(1, 8,
+        "hpx::parallel::rotate is deprecated, use hpx::ranges::rotate "
+        "instead") typename util::detail::algorithm_result<ExPolicy,
+        util::in_out_result<typename hpx::traits::range_iterator<Rng>::type,
+            typename hpx::traits::range_iterator<Rng>::type>>::type
+        rotate(ExPolicy&& policy, Rng&& rng,
+            typename hpx::traits::range_iterator<Rng>::type middle)
+    {
+        return rotate(std::forward<ExPolicy>(policy), hpx::util::begin(rng),
+            middle, hpx::util::end(rng));
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // clang-format off
+    template <typename ExPolicy, typename Rng, typename OutIter,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::is_execution_policy<ExPolicy>::value && hpx::traits::is_range<
+                Rng>::value && hpx::traits::is_iterator<OutIter>::value)>
+    // clang-format on
+    HPX_DEPRECATED_V(1, 8,
+        "hpx::parallel::rotate_copy is deprecated, use "
+        "hpx::ranges::rotate_copy instead")
+        typename util::detail::algorithm_result<ExPolicy,
+            util::in_out_result<typename hpx::traits::range_iterator<Rng>::type,
+                OutIter>>::type rotate_copy(ExPolicy&& policy, Rng&& rng,
+            typename hpx::traits::range_iterator<Rng>::type middle,
+            OutIter dest_first)
     {
         return rotate_copy(std::forward<ExPolicy>(policy),
             hpx::util::begin(rng), middle, hpx::util::end(rng), dest_first);
     }
 }}}    // namespace hpx::parallel::v1
+
+namespace hpx { namespace ranges {
+    template <typename I, typename O>
+    using rotate_copy_result = hpx::parallel::util::in_out_result<I, O>;
+
+    ///////////////////////////////////////////////////////////////////////////
+    // DPO for hpx::ranges::rotate
+    HPX_INLINE_CONSTEXPR_VARIABLE struct rotate_t final
+      : hpx::functional::tag_fallback<rotate_t>
+    {
+    private:
+        // clang-format off
+        template <typename FwdIter, typename Sent,
+          HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_forward_iterator<FwdIter>::value &&
+                hpx::traits::is_sentinel_for<Sent, FwdIter>::value
+          )>
+        // clang-format on
+        friend subrange_t<FwdIter, Sent> tag_fallback_dispatch(
+            hpx::ranges::rotate_t, FwdIter first, FwdIter middle, Sent last)
+        {
+            static_assert(hpx::traits::is_forward_iterator<FwdIter>::value,
+                "Requires at least forward iterator.");
+
+            return hpx::parallel::util::get_subrange<FwdIter, Sent>(
+                hpx::parallel::v1::detail::rotate<FwdIter>().call(
```

Comment:
It should be hpx::parallel::v1::detail::rotate<util::in_out_result<FwdIter, Sent>>(), we use this to specifiy what the return type would be

Reply:
fixed. Thanks!